### PR TITLE
8372 gdcc xoai library

### LINKED ISSUE #8372: Update and reorganize the XOAI dependencies under local_lib
Currently, Dataverse codebase uses a custom patched XOAI 4.1.0, provided in `/local_lib`.

It would be a good idea to

1. Find out what patches have been applied (needs anatomization of patched JAR sources with original sources)
2. Fork https://github.com/DSpace/xoai (which is the moved repo from https://github.com/lyncode) to @gdcc
3. Find out if our patches are still needed
4. As development in https://github.com/DSpace/xoai has stalled (see e.g. https://github.com/DSpace/xoai/issues/72#issuecomment-557292929), make it our package
5. Update LOTS of dependencies and especially get rid of log4j-1.2!
6. Setup Github Workflow, Code Coverage etc.
7. Make releases to Maven Central
8. Use & test in Dataverse upstream code

It might be worth a try to 1) not rename the packages but still publish under our @gdcc Maven group id and 2) create fork and a pull request for upstream plus setting up wei/pull to auto-create pull requests when DSpace updates their branch.

### PR DESCRIPTION
**Pull request by landreev**
**What this PR does / why we need it**:
Big technical debt PR; eliminates the custom jars of the old lyncode XOAI library; replaces them with dependencies for the new io.gdcc.xoai packages; fixes some known issues in the process. 
Still a draft; basic functionality is there; still working on more tests, although I'm considering splitting that out as a separate issue. 

**Which issue(s) this PR closes**:

- Closes #8372

**Special notes for your reviewer**:

The scope of this PR has ballooned A LOT. It was initially started under the assumption that it was simply to accommodate the move of the XOAI libraries from local_lib into a gdcc repo _as they were_, with only some minimal changes, leaving any further improvements and cleanup of the modifications in the original fork for later, as separate issues/PRs. Instead a very thorough refactoring of the entire xoai package was done before a stable 5.0 RC release was made on gdcc/xoai, reimplementing and/or getting rid of the early dataverse hacks and solving many issues in the process. Huge thanks to @poikilotherm for all the work there. 
This is all a "good thing, not a bad thing"; it was all very real technical debt that we needed to address rather sooner than later. This has allowed to very much streamline the dataverse-side OAI implementation and get rid of most of the code that was maintained there. There's more that still needs to be done on the dataverse side for sure. But since this has already been a much larger effort than was originally scheduled, I really want to move the PR along and then proceed to handle the remaining improvements as separate tasks. 

**Suggestions on how to test this**:

**Does this PR introduce a user interface change? If mockups are available, please link/include them here**:

**Is there a release notes update needed for this change?**:

**Additional documentation**:

### CODE REVIEW

**Comment by landreev** [comment]
@poikilotherm Hi, Harvard just gave us all an extra mini-vacation, but I'm back at work. I made this draft PR last week that finally gets rid of the old custom jars in local_lib. Once again, I cannot fully express my gratitude for your leading of this effort.  
I was able to easily get rid of the extra xoai classes on the dataverse side; and to implement almost everything using the new mechanisms provided in gdcc.xoai, as we planned. I created a local `Condition` in order to pass the MetadataFormat to the `getItems()` methods in the ItemRepository implementation via a ScopedFilter. Etc. etc. 
There are extra things that need to be cleaned up: the `Date -> Instant` change is currently handled via quick conversions between the two (i.e., the underlying entities are left unchanged, using `Date` for the stamps); it clearly needs more API tests. I'm working on this, but considering splitting the task into a separate issue. 

There is one thing I could not implement 1:1 as it was: the handling of the proprietary "Dataverse JSON harvesting". I wanted to consult with you about this. The way the JSON api line has been encoded in the original implementation: 
```
<metadata directApiCall="..."/>
```
 \- This doesn't seem to be possible, to implement this using the gdcc.xoai framework without adding ugly hacks on the library side again (?). And it was a somewhat unfortunate choice in the first place. I feel like *if* we want to keep this "json harvesting" going forward, it should be something like
```
<metadata>
   <dataverseCustom directApiCall="..."/>
</metadata>
``` 
instead. This I was able to easily implement in the item repository; but it obviously introduces a backward incompatibility.  Do you have any thoughts on this? I'm also going to talk to the core dev. team, about whether we want to continue supporting this functionality; and/or if we want to change it.

File: pom.xml
Comment (poikilotherm) [inline pom.xml:390]: Not sure it's necessary to include the common dep here, as it is already a dep of the data & service provider.

Yeah, I know, one should add all transitive deps here that one is using within its own code, but this is all the same library...

File: pom.xml
Comment (poikilotherm) [inline pom.xml:394]: Would it make sense to release a BOM for this lib and use it in `<dependencyManagement>` to avoid these repeats here? (Following [JLBP-4](http://java.jonathangiles.net/JLBP-4))

File: src/main/java/edu/harvard/iq/dataverse/harvest/client/HarvesterServiceBean.java
Comment (poikilotherm) [inline src/main/java/edu/harvard/iq/dataverse/harvest/client/HarvesterServiceBean.java:392]: IMHO this code is really hacky Java with C origin, also using some Exception anti-patterns.

May I suggest starting to use try-with-resources, considering [`InputStream.transferTo()`](https://docs.oracle.com/en/java/javase/11/docs/api/java.base/java/io/InputStream.html#transferTo(java.io.OutputStream)) and looking at the modern [Java included HTTP client](https://docs.oracle.com/en/java/javase/11/docs/api/java.net.http/java/net/http/package-summary.html)? (See also https://www.mocklab.io/blog/which-java-http-client-should-i-use-in-2020/)

File: src/main/java/edu/harvard/iq/dataverse/harvest/client/HarvesterServiceBean.java
Comment (poikilotherm) [inline src/main/java/edu/harvard/iq/dataverse/harvest/client/HarvesterServiceBean.java:410]: If you don't expect the pattern to occur multiple times, you might want to avoid this. It's very slow compared to the other available options.

File: src/main/java/edu/harvard/iq/dataverse/harvest/server/OAIRecordServiceBean.java
Comment (poikilotherm) [inline src/main/java/edu/harvard/iq/dataverse/harvest/server/OAIRecordServiceBean.java:356]: **suggestion:** Please not the clock skewing is [done by the XOAI lib already](https://github.com/gdcc/xoai/blob/88b06ef3cb401373b158cb9b71b57286b59db2a3/xoai-data-provider/src/main/java/io/gdcc/xoai/dataprovider/request/RequestBuilder.java#L239).

You can alter the skewing behavior by overriding the `RepositoryConfiguration.skewUntil(()` method if necessary.

File: src/main/java/edu/harvard/iq/dataverse/harvest/server/web/servlet/OAIServlet.java
Comment (poikilotherm) [inline src/main/java/edu/harvard/iq/dataverse/harvest/server/web/servlet/OAIServlet.java:244]: **suggestion:** You might consider creating the writer within a try-with-resources to avoid any spilled beans.

File: src/main/java/edu/harvard/iq/dataverse/harvest/server/web/servlet/OAIServlet.java
Comment (poikilotherm) [inline src/main/java/edu/harvard/iq/dataverse/harvest/server/web/servlet/OAIServlet.java:255]: **nitpick**: I'm not sure if this might be an Exception anti-pattern here.

1) Is the ServletException exposed to a user, which will include the pesky internal details? (That would be an anti-pattern)
2) Is the ServletException logged by the app server, too? In that case: would it be wise to avoid double logging the same thing? (Which would also be an anti-pattern)

File: src/main/java/edu/harvard/iq/dataverse/harvest/server/web/servlet/OAIServlet.java
Comment (poikilotherm) [inline src/main/java/edu/harvard/iq/dataverse/harvest/server/web/servlet/OAIServlet.java:379]: **praise:** Yeah! Away with that! :partying_face:

File: src/main/java/edu/harvard/iq/dataverse/harvest/server/xoai/DataverseXoaiItemRepository.java
Comment (poikilotherm) [inline src/main/java/edu/harvard/iq/dataverse/harvest/server/xoai/DataverseXoaiItemRepository.java:60]: **issue:** Please don't do this. There is exactly one valid use case for this: when a [`ListMetadataFormats`](http://www.openarchives.org/OAI/openarchivesprotocol.html#ListMetadataFormats) request is sent, it might carry an identifier to retrieve all available metadata formats for this specific item.

Please be aware that you extended the method signature: you provide a return type "Item", whereas the interface only request an "ItemIdentifier" (which is inherited by "Item", so this works).

The available metadata formats for the item are provided via the `Context`. It will default to add all formats the context knows about, but you can add more specific things if necessary.

I'd strongly advise against throwing the exception as is now.

File: src/main/java/edu/harvard/iq/dataverse/settings/SettingsServiceBean.java
Comment (poikilotherm) [inline src/main/java/edu/harvard/iq/dataverse/settings/SettingsServiceBean.java:539]: **nitpick:** Please make the writing consistent and use a capital letter within the `Key` enum.

**though:** Can we pretty please avoid adding even more database settings for things that definitely are static things? #7000 got pretty far already.

File: src/main/java/edu/harvard/iq/dataverse/util/SystemConfig.java
Comment (poikilotherm) [inline src/main/java/edu/harvard/iq/dataverse/util/SystemConfig.java:1233]: **though:** can we please use #7000 for this?

File: src/main/java/edu/harvard/iq/dataverse/harvest/server/OAIRecordServiceBean.java
Comment (landreev) [reply src/main/java/edu/harvard/iq/dataverse/harvest/server/OAIRecordServiceBean.java:356]: Sure, we discussed this of course, I forgot to remove it.

File: src/main/java/edu/harvard/iq/dataverse/harvest/server/xoai/DataverseXoaiItemRepository.java
Comment (landreev) [reply src/main/java/edu/harvard/iq/dataverse/harvest/server/xoai/DataverseXoaiItemRepository.java:60]: I didn't think/remember about this use case (ListMetadataFormats with the identifier parameter). Makes sense.
(Why is the method called getItem() though; and not getItemIdentifier(), for consistency with the other method?)

File: src/main/java/edu/harvard/iq/dataverse/harvest/server/web/servlet/OAIServlet.java
Comment (poikilotherm) [inline src/main/java/edu/harvard/iq/dataverse/harvest/server/web/servlet/OAIServlet.java:84]: **praise:** All hail to mighty MPCONFIG!

(Might be refactored with what is in #8823, but it's a great start!!! :partying_face: )

File: src/main/java/edu/harvard/iq/dataverse/harvest/server/xoai/DataverseXoaiItemRepository.java
Comment (landreev) [reply src/main/java/edu/harvard/iq/dataverse/harvest/server/xoai/DataverseXoaiItemRepository.java]: Sure, will do.

File: pom.xml
Comment (landreev) [reply pom.xml:389]: This one I'm not sure about honestly, if it's worth bothering.

File: src/main/java/edu/harvard/iq/dataverse/harvest/server/xoai/DataverseXoaiItem.java
Comment (poikilotherm) [inline src/main/java/edu/harvard/iq/dataverse/harvest/server/xoai/DataverseXoaiItem.java:23]: **nitpick:** shouldn't this class be `final`? (Yes, I'm a huge fan of non-modifiable by default)

File: src/main/java/edu/harvard/iq/dataverse/harvest/server/xoai/DataverseXoaiItem.java
Comment (poikilotherm) [inline src/main/java/edu/harvard/iq/dataverse/harvest/server/xoai/DataverseXoaiItem.java:60]: **nitpick:** Shouldn't this return an unmodifiable list to avoid changing the objects? (e. g. via `Collections.unmodifiableList()`)

File: src/main/java/edu/harvard/iq/dataverse/harvest/server/xoai/DataverseXoaiItem.java
Comment (poikilotherm) [inline src/main/java/edu/harvard/iq/dataverse/harvest/server/xoai/DataverseXoaiItem.java:32]: **thought:** Maybe move this to the declaration + make the list `final`?

File: src/main/java/edu/harvard/iq/dataverse/harvest/server/xoai/DataverseXoaiSetRepository.java
Comment (poikilotherm) [inline src/main/java/edu/harvard/iq/dataverse/harvest/server/xoai/DataverseXoaiSetRepository.java:19]: **nitpick:** Should be final.

File: src/main/java/edu/harvard/iq/dataverse/harvest/server/xoai/DataverseXoaiItemRepository.java
Comment (poikilotherm) [inline src/main/java/edu/harvard/iq/dataverse/harvest/server/xoai/DataverseXoaiItemRepository.java:43]:
```suggestion
    private final OAIRecordServiceBean recordService;
    private final DatasetServiceBean datasetService;
    private final String serverUrl; 
```

File: src/main/java/edu/harvard/iq/dataverse/harvest/server/xoai/DataverseXoaiItemRepository.java
Comment (poikilotherm) [inline src/main/java/edu/harvard/iq/dataverse/harvest/server/xoai/DataverseXoaiItemRepository.java:52]:
```suggestion
    private final List<DataverseXoaiItem> list = new ArrayList<>();
```

File: src/main/java/edu/harvard/iq/dataverse/harvest/server/xoai/DataverseXoaiItemRepository.java
Comment (poikilotherm) [inline src/main/java/edu/harvard/iq/dataverse/harvest/server/xoai/DataverseXoaiItemRepository.java:70]:
```suggestion
                    return new DataverseXoaiItem(oaiRecord);
```

File: src/main/java/edu/harvard/iq/dataverse/harvest/server/xoai/DataverseXoaiItemRepository.java
Comment (poikilotherm) [inline src/main/java/edu/harvard/iq/dataverse/harvest/server/xoai/DataverseXoaiItemRepository.java:65]:
```suggestion
```

File: src/main/java/edu/harvard/iq/dataverse/harvest/server/xoai/DataverseXoaiItemRepository.java
Comment (poikilotherm) [inline src/main/java/edu/harvard/iq/dataverse/harvest/server/xoai/DataverseXoaiItemRepository.java:114]: **suggestion:** Why not simple throw the exception here to make it more explicit and reduce the danger of missing this when refactoring? (Is there a unit test for the behavior?)

File: src/main/java/edu/harvard/iq/dataverse/harvest/server/xoai/DataverseXoaiItemRepository.java
Comment (poikilotherm) [inline src/main/java/edu/harvard/iq/dataverse/harvest/server/xoai/DataverseXoaiItemRepository.java:133]: **suggestion:** This very much belongs together - shouldn't all this be inside the `try` block? (Which also could throw directly, as mentioned before/above)

File: src/main/java/edu/harvard/iq/dataverse/harvest/server/xoai/DataverseXoaiItem.java
Comment (landreev) [reply src/main/java/edu/harvard/iq/dataverse/harvest/server/xoai/DataverseXoaiItem.java:60]: Sure. Or I could just delete these method and rely on the empty list implementation in Item.java; since we have no use for it yet.

Comment (landreev) [reply src/main/java/edu/harvard/iq/dataverse/harvest/server/xoai/DataverseXoaiItem.java:60]: Making the returned list unmodifiable will in fact make it impossible to add any Abouts to the item; since there are no withAbout() or setAbout() methods yet. Yes, I feel like just removing both the variable and the method from the class, until we have some need/use case for this "about" thing.

File: src/main/java/edu/harvard/iq/dataverse/harvest/server/xoai/DataverseXoaiItemRepository.java
Comment (landreev) [reply src/main/java/edu/harvard/iq/dataverse/harvest/server/xoai/DataverseXoaiItemRepository.java:51]: I have a better solution.

File: src/main/java/edu/harvard/iq/dataverse/harvest/server/xoai/DataverseXoaiItemRepository.java
Comment (landreev) [reply src/main/java/edu/harvard/iq/dataverse/harvest/server/xoai/DataverseXoaiItemRepository.java:82]: This could be a matter of taste. I feel like the way this is implemented here makes sense. This loop logic is not only about waiting to create the item until we know that there's at least one record available; it is also to add the extra sets to the item that was created and populated with metadata when the first record was encountered, when there are multiple records.
(there are multiple records in the database per identifier, one record for every set in which it occurs)

File: src/main/java/edu/harvard/iq/dataverse/harvest/server/xoai/DataverseXoaiItemRepository.java
Comment (landreev) [reply src/main/java/edu/harvard/iq/dataverse/harvest/server/xoai/DataverseXoaiItemRepository.java:127]: I will move the code duplicated between this method and getItems(), and this will be addressed in the process.

File: src/main/java/edu/harvard/iq/dataverse/harvest/server/xoai/DataverseXoaiItemRepository.java
Comment (landreev) [reply src/main/java/edu/harvard/iq/dataverse/harvest/server/xoai/DataverseXoaiItemRepository.java:111]: ok.

**Review by poikilotherm — APPROVED** [review]
Looks good to me. Thanks for incorporating all those changes!

@landreev Shall I bake a 5.0.0-RC2 or do you want me to go "final" directly?

**Comment by kcondon** [comment]
Issues so far:
1. Datacite, oai_datacite, and dataverse_json formats do not work. These also do not work in the prior version so may be out of scope. Datacite, oai_datacite log that those formats are not supported for import and there is an open issue to implement. The dataverse_json issue is unclear at the moment but it had worked in the original implementation with some isolated failing records but currently fails in existing implementation, not sure when it stopped working but could look at licensing.

**Comment by landreev** [comment]
As for the first 2 formats under 1., we really don't support them (at least for now), so we should probably hide them from the pulldown and/or not allow people to select any formats except the supported ones. 
As for dataverse_json, just to confirm, are you saying that it's failing with the same license issue on develop/5.11 as well, not just on my branch?

**Comment by landreev** [comment]
As for 
> 2. Harvesting as a server using a v5.11 client, same formats fail but json reports a different error: unable to find valid certification path to requested target

dataverse_json does require an additional http call (directly to the export API, to obtain json), outside of the OAI library. So it does not surprise me that the code that makes this call may be more sensitive to whether the cert is valid or not. But if the client is 5.11 in this context, then this is the existing behavior.

**Comment by kcondon** [comment]
@landreev I wasn't saying that but since you asked I have now tested json using v5.11 as both client (demo) and server (prod) and get the same license error.

**Comment by poikilotherm** [comment]
@kcondon sounds like this reveals some internal problems?! If there is anything I can do, please reach out (here/Matrix/Slack/Issue in XOAI/...)

**Comment by landreev** [comment]
@poikilotherm 
> sounds like this reveals some internal network problems?! If there is anything I can do, please reach out (here/Matrix/Slack/Issue in XOAI/...)

I don't think it's network problems. It looks like it literally is what it says - a problem trying to validate the cert of the remote server. Which in turn could have happened for various reasons... I'm working with Kevin on taking these cert issues out of the equation, so that we can focus on testing the XOAI functionality proper. 
Regardless, if we need anything from you, we absolutely know where you live, yes!

**Comment by landreev** [comment]
> I wasn't saying that but since you asked I have now tested json using v5.11 as both client (demo) and server (prod) and get the same license error.

OK, good to know. I'll figure this out, even if it's not specific to my branch. Again, to be able to take this error out of the equation/be able to test before vs. after. 
We were having various license-related import issues right after that PR that refactored everything license-related; with somewhat similar errors in the exception stack traces. Most of those were trivial - like, there were places in the code where the license server bean was not properly passed to the json parser (the new requirement). Although I thought we had fixed them all. Anyway,  I'll figure it out, will talk to Jim if necessary.

**Comment by landreev** [comment]
So, about those dataverse_json import errors: 
The `Invalid license: CC0` error was there because some of the prod. datasets in the set you were testing did not get re-exported after 5.10, when the license was changed. All the datasets have now been re-exported, so this should no longer be an issue.
However, *every single* prod. dataset is going to be failing to import/harvest in dataverse_json by 5.11 clients now, with a different error: `Specified metadatalanguage not allowed`. This is on account of #8868. 

I'm guessing this means that the only way to test harvesting dataverse_json now is to have the new `metadataLanguages` setting defined on both the server and client; and have every dataset in the OAI set configured with a `metadataLanguage` value recognized by the client... which frankly feels like serious pain. So waiting to have a fix for #8868 merged may be a better idea. 

On the [somewhat] plus side, I still think that the very fact that these dataverse_json harvests are failing with the same import errors in my branch does mean that the harvesting framework is working as it should be. By the nature of the changes there, if it were broken, it would be failing to obtain the json, and never getting to the import stage.

**Comment by landreev** [comment]
(Just put my name on the PR to update the status, as requested in standup)
The last comment above is still an accurate description of the status of the PR: the serious issues with harvesting JSON  that were identified during the first QA pass are unrelated to the PR itself. The issues in question should be addressed in #8868, which appears to be close to being done. Once it's merged, this PR can be re-tested. To the best of my knowledge everything else in it should be working; at least working as well, or better than it has been up until now, which is the goal of this PR. (for example, one small thing identified above - that we allow an admin to select a format from the pulldown that Dataverse does not know how to import - is a pre-existing issue; that I'm assuming we don't want to try to address during this pass).

**Comment by landreev** [comment]
Since the json export PR has been merged, this issue is ready to be re-tested. 
Re-testing the functionality directly involving json export (i.e., harvesting in dataverse_json format) is fairly straightforward (just needs going through all the combinations of "old" and "new" on the client and server ends). But then I'm not sure if whoever ends up working on wrapping up this PR will want to re-test *everything* from scratch (?). 
I will work with this person to provide guidance/answer questions.

**Comment by landreev** [comment]
OK, everything has been re-tested very carefully. 
Additionally, harvesting dataverse_json forrmat, that could not be fully tested during the first QA attempt mainly because of the export bug in 5.11 was thoroughly tested as follows: 

This branch simplified the process of harvesting of our proprietary json records; it skips the step where the client gets a hacked GetRecord entry from the server and extracts the url for obtaining the JSON record. It just calls the export API directly, since the url for it can be derived from knowing the server address. However, any remaining Dataverse clients not yet upgraded to this version will continue using the old model. So for backward compatibility, this new version of xoai-server still knows how to produce the old-style, custom version of GetRecord with the url embedded (which is unfortunate, because it's really hacky. But we will get rid of it in a future version, once it is safe to assume that few if any old-style clients harvesting this format are still out there). So all three possible combinations of harvesting dataverse_json had to be tested: new-to-new; old-to-new; new-to-old. Testing was done using my own dev. build as the client, and either dataverse-internal or the perf. cluster as the server.

**Pull request merged**
